avocado/plugins/diff.py

```diff
@@ -0,0 +1,421 @@
+from avocado.core.settings import settings
+
+
+LOG = logging.getLogger("avocado.app")
+
+
+class Diff(CLICmd):
+
+    """
+    Implements the avocado 'diff' subcommand
+    """
+
+    name = 'diff'
+    description = 'Shows the difference between 2 jobs.'
+
+    def __init__(self):
+        self.term = output.TERM_SUPPORT
+        self.std_diff_output = True
+
+    def configure(self, parser):
+        """
+        Add the subparser for the diff action.
+
+        :param parser: Main test runner parser.
+        """
+        parser = super(Diff, self).configure(parser)
+
+        parser.add_argument("jobids",
+                            default=[], nargs=2,
+                            metavar="<JOB>",
+                            help='A job reference, identified by a (partial) '
+                            'unique ID (SHA1) or test results directory.')
+
+        parser.add_argument('--html', type=str,
+                            metavar='FILE',
+                            help='Enable HTML output to the FILE where the '
+                            'result should be written.')
+
+        parser.add_argument('--open-browser',
+                            action='store_true',
+                            default=False,
+                            help='Generate and open a HTML report in your '
+                            'preferred browser. If no --html file is '
+                            'provided, create a temporary file.')
+
+        parser.add_argument('--diff-filter',
+                            dest='diff_filter',
+                            type=self._validate_filters,
+                            default=['all'],
+                            help='Comma separated filter of diff sections: '
+                            '(no)cmdline,(no)time,(no)variants,(no)results,\n'
+                            '(no)config,(no)sysinfo (defaults to all '
+                            'enabled).')
+
+        parser.add_argument('--paginator',
+                            choices=('on', 'off'), default='on',
+                            help='Turn the paginator on/off. '
+                            'Current: %(default)s')
+
+        parser.add_argument('--create-reports', action='store_true',
+                            help='Create temporary files with job reports '
+                            '(to be used by other diff tools)')
+
+        parser.epilog = 'By default, a textual diff report is generated '\
+                        'in the standard output.'
+
+    def run(self, args):
+        job1_dir, job1_id = self._setup_job(args.jobids[0])
+        job2_dir, job2_id = self._setup_job(args.jobids[1])
+
+        job1_data = self._get_job_data(job1_dir)
+        job2_data = self._get_job_data(job2_dir)
+
+        report_header = 'Avocado Job Report\n'
+        job1_results = [report_header]
+        job2_results = [report_header]
+
+        if 'all' in args.diff_filter or 'cmdline' in args.diff_filter:
```

Comment:
well the main point of the `_validate_filters` enhancement was to get rid of `all`. It's not needed, the only remnant of it is the `add_argument`, where the default should be the list of all sections and not `all`, because the `default` is not processed by the `type` function.